### PR TITLE
Only ONE StorageClass can be marked as Default

### PR DESCRIPTION
**Pull request by Ranger-X**
## Description
Implementation of a unified mechanism for specifying default `StorageClass`.

- Add the `defaultClusterStorageClass` setting to `global` module config.
- Deckhouse uses this setting to determine whether the default `StorageClass` needs to be updated. If an update is needed:
  - Removes the `storageclass.kubernetes.io/is-default-class` annotations (if any) from all `StorageClass`.
  - Sets the `storageclass.kubernetes.io/is-default-class=true` annotation on the `StorageClass` specified in `global.defaultClusterStorageClass`.
  - Prevent adding/changing/removing the `storageclass.kubernetes.io/is-default-class` annotation on `StorageClass` resources for any users except the user under which Deckhouse is running

## Why do we need it, and what problem does it solve?
- There is currently no single mechanism (config value) for specifying a default `StorageClass` in Deckhouse platform.
- In Kubernetes it is possible to assign the annotation `storageclass.kubernetes.io/is-default-class=true` to multiple `StorageClass`. 
Quote from [k8s documentation](https://kubernetes.io/docs/concepts/storage/storage-classes/#default-storageclass):
  > If you set the `storageclass.kubernetes.io/is-default-class` annotation to true on more than one StorageClass in your cluster, and you then create a PersistentVolumeClaim with no storageClassName set, Kubernetes uses the most recently created default StorageClass.

## What is the expected result?
The user can configure the default `StorageClass` via the `defaultClusterStorageClass` parameter in `global` module config.

## Info for reviewers
Manual tests were performed (and passed) on a cluster in Yandex.Cloud.

2 cases were tested (including deleting these tests clusters via dhctl):
1. When updating a cluster from the current `main` branch to this PR
2. When bootstrapping a new cluster from this PR version

Also note, that almost all E2E tests passed, except [EKS](https://github.com/deckhouse/deckhouse/actions/runs/11511947811/job/32046185525) with error due to reasons beyond the control of this PR.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: global-hooks
type: feature
summary: Add the `global.defaultClusterStorageClass` setting.
impact: cloud-provider's `storageClass.default` parameter was deprecated (not used anymore) and replaced with `global.defaultClusterStorageClass`
impact_level: default
```

### CODE REVIEW

File: global-hooks/migrate/migrate_global_storage_class.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 Flant JSC
```
Comment (AleksZimin) [inline global-hooks/migrate/migrate_global_storage_class.go:2]:
```suggestion
Copyright 2024 Flant JSC
```

**Review by AleksZimin — CHANGES_REQUESTED** [review]

File: global-hooks/default_cluster_storage_class.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+	annotToCheck := []string{
+		"storageclass.beta.kubernetes.io/is-default-class",
+		"storageclass.kubernetes.io/is-default-class",
+	}
+
+	isDefault := false
+	for _, annot := range annotToCheck {
+		if v, ok := annotations[annot]; ok && strings.ToLower(v) == "true" {
+			isDefault = true
```
Comment (astef) [inline global-hooks/default_cluster_storage_class.go:134]: or just `return true` ?

File: global-hooks/default_cluster_storage_class.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		} else {
+			if isMarkedDefault(&sc) {
+				// we must remove default-annotation from this StorageClass because only one StorageClass (which name in defaultClusterStorageClass) can be default
+				input.LogEntry.Warnf("Remove default annotations from storage class %q", sc.GetName())
+
+				patch := map[string]any{
+					"metadata": map[string]any{
+						"annotations": map[string]any{
+							"storageclass.beta.kubernetes.io/is-default-class": nil,
+							"storageclass.kubernetes.io/is-default-class":      nil,
```
Comment (astef) [inline global-hooks/default_cluster_storage_class.go:110]: May be extract such keys into constants? They're used multiple times in this file.

File: go_lib/hooks/storage_class_change/hook.go [hunk 4]
```diff
@@ -196,9 +196,15 @@ func calculateEffectiveStorageClass(input *go_hook.HookInput, args Args, current
 	}
 
 	if input.ConfigValues.Exists("global.storageClass") {
```
Comment (Ranger-X) [inline go_lib/hooks/storage_class_change/hook.go:198]: remove and try tests

File: global-hooks/default_cluster_storage_class.go (hunk 2 above)
Comment (Ranger-X) [reply global-hooks/default_cluster_storage_class.go]: Yeap. You are right :-) Done

File: global-hooks/default_cluster_storage_class.go (hunk 3 above)
Comment (Ranger-X) [reply global-hooks/default_cluster_storage_class.go]: Fixed

File: ee/modules/030-cloud-provider-zvirt/hooks/discover.go [hunk 5]
```diff
@@ -222,23 +222,11 @@ func getStorageClassName(value string) string {
 }
 
 func setStorageClassesValues(
-	values *go_hook.PatchableValues,
+	input *go_hook.HookInput,
 	storageClasses []storageClass,
-	defaultSCName string,
+	_ string,
```
Comment (Ranger-X) [inline ee/modules/030-cloud-provider-zvirt/hooks/discover.go:227]: fix it

File: go_lib/hooks/storage_class/hook.go
Comment (Ranger-X) [inline go_lib/hooks/storage_class/hook.go:1]: try use this hook for all other cloud-providers

File: ee/modules/030-cloud-provider-vcd/hooks/discover.go [hunk 6]
```diff
@@ -175,23 +175,11 @@ func handleDiscoveryDataVolumeTypes(input *go_hook.HookInput, volumeTypes []v1al
...
-		values.Set("cloudProviderVcd.internal.defaultStorageClass", def.String())
-		return
-	}
-
-	if defaultSCName != "" {
-		values.Set("cloudProviderVcd.internal.defaultStorageClass", defaultSCName)
-		return
-	}
-	values.Remove("cloudProviderVcd.internal.defaultStorageClass")
+func setStorageClassesValues(input *go_hook.HookInput, storageClasses []storageClass, _ string) {
```
Comment (Ranger-X) [inline ee/modules/030-cloud-provider-vcd/hooks/discover.go:181]: fix it

File: helm_lib/charts/deckhouse_lib_helm/templates/_module_storage_class.tpl
Comment (Ranger-X) [inline helm_lib/charts/deckhouse_lib_helm/templates/_module_storage_class.tpl:1]: We must accept [PR in lib-helm](https://github.com/deckhouse/lib-helm/pull/71/files) and make update-lib-helm before merge this one

File: global-hooks/openapi/doc-ru-config-values.yaml [hunk 7]
```diff
@@ -6,6 +6,13 @@ properties:
         * Если значение не указано, используется автоматически определяемый `global.discovery.defaultStorageClass`. Если `global.discovery.defaultStorageClass` не определен, используется `emptyDir`.
         * Этот параметр имеет смысл использовать только в исключительных ситуациях.
         * Этот параметр применяется в процессе включения модуля.
+  defaultClusterStorageClass:
+    type: string
+    default: null
+    description: |
+      Имя storage class, который будет использоваться для всего кластера.
+        * Если значение указано, то все вновь создаваемые PVC будут использовать указанный storage class.
+        * Этот параметр применяется в процессе включения модуля.
```
Comment (Ranger-X) [inline global-hooks/openapi/doc-ru-config-values.yaml:15]:
```suggestion
      Имя storage class, который будет использоваться в кластере по умолчанию.
        * Если значение НЕ указано, то будет сохранен уже существующий StorageClass по умолчанию, если он есть в кластере. Иначе будет использован первый StorageClass из создаваемых модулем cloud-provider (если какой-либо модуль cloud-provider-а используется).
        * Используйте этот параметр, если хотите поменять StorageClass по умолчанию в кластере, т.к. "ручное" добавление default-аннотации даже к собственному StorageClass запрещено.
```

**Review by z9r5 — APPROVED** [review]

File: modules/030-cloud-provider-yandex/docs/FAQ.md [hunk 8]
```diff
@@ -47,19 +47,12 @@ If the dhcpOptions parameter is set, all DNS are routed to the DNS servers speci
 
 ## How to set a custom StorageClass as default?
 
-Do the following to set a custom StorageClass as default:
+Specify the StorageClass name in the [defaultClusterStorageClass](../../deckhouse-configure-global.html#parameters-defaultclusterstorageclass) parameter in the `global` module settings.
+Note that after doing so, the `storageclass.kubernetes.io/is-default-class='true'` annotation will be removed from the StorageClass that was previously be as the default one.
```
Comment (miklezzzz) [inline modules/030-cloud-provider-yandex/docs/FAQ.md:51]:
```suggestion
Note that after doing so, the `storageclass.kubernetes.io/is-default-class='true'` annotation will be removed from the StorageClass that was previously set as the default one.
```

**Review by miklezzzz — APPROVED** [review]

File: modules/030-cloud-provider-yandex/openapi/config-values.yaml
Comment (Ranger-X) [inline modules/030-cloud-provider-yandex/openapi/config-values.yaml:1]: Remove `defaultStorageClass` from values.yaml

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @aleksey-su. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/vcd'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/eks'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/eks'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/aws'...

**Review by aleksey-su — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/eks'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/vsphere'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/vcd'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/azure'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Review by nabokihms — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/yandex-cloud'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @Ranger-X. I've started the workflow for label 'e2e/run/aws'...

**Review by AleksZimin — APPROVED** [review]

**Pull request merged**